storage/src/adb/any/fixed/mod.rs

```diff
@@ -256,3 +261,133 @@ where
 
     Ok((proof, ops))
 }
+
+/// Find and return the location of the update operation for `key`, if it exists. The cursor is
+/// positioned at the matching location, and can be used to update or delete the key.
+async fn find_update_op<E, C, K, O>(
+    log: &Journal<E, O>,
+    cursor: &mut C,
+    key: &K,
+) -> Result<Option<Location>, Error>
+where
+    E: Storage + Clock + Metrics,
+    C: Cursor<Value = Location>,
+    K: Array,
+    O: FixedOperation<Key = K>,
+{
+    while let Some(&loc) = cursor.next() {
+        let op = log.read(*loc).await?;
+        let k = op.key().expect("operation without key");
+        if *k == *key {
+            return Ok(Some(loc));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Append `op` to the log and add it to the MMR. The operation will be subject to rollback
+/// until the next successful `commit`.
+async fn apply_op<E, O, H>(
+    mmr: &mut Mmr<E, H>,
+    log: &mut Journal<E, O>,
+    hasher: &mut Standard<H>,
+    op: O,
+) -> Result<(), Error>
+where
+    E: Storage + Clock + Metrics,
+    O: FixedOperation,
+    H: CHasher,
+{
+    let encoded_op = op.encode();
+
+    // Append operation to the log and update the MMR in parallel.
+    try_join!(
+        mmr.add_batched(hasher, &encoded_op).map_err(Error::Mmr),
+        log.append(op).map_err(Error::Journal)
+    )?;
+
+    Ok(())
+}
+
+// Moves the given operation to the tip of the log if it is active, rendering its old location
+// inactive. If the operation was not active, then this is a no-op. Returns the old location
+// of the operation if it was active.
+pub(crate) async fn move_op_if_active<E, I, O, H>(
+    snapshot: &mut I,
+    mmr: &mut Mmr<E, H>,
+    log: &mut Journal<E, O>,
+    hasher: &mut Standard<H>,
+    op: O,
+    tip_loc: Location,
+    old_loc: Location,
```

Comment:
the large number of arguments here is kind of annoying... could wrap in a struct I suppose?